### ISSUE TITLE
Swift4 Codable support

### ISSUE THREAD

**Issue by Bersaelor** [apple/swift-protobuf#657]
Maybe thats a stupid question, as I haven't used protobuf yet,

but are there any plans to support the Swift4 `Codable` protocol?

I could use a fast alternative to `JSONEncoder` or `PropertyListEncoder` and it feels like protobuf may be able to offer a similar API as those two for serialization?

**Comment by thomasvl** [comment]
Protobuf includes its own [spec for JSON support](https://developers.google.com/protocol-buffers/docs/proto3#json), which doesn't always line up with what folks want. The protobuf one uses strings for some numeric types because it strives for interop, and some implementation (javascript), can have issues with larger numerical values.

So at the moment, we have a json interface exposed, it isn't `Codable`, but adding a second seems like it would be somewhat confusing.

**Comment by Bersaelor** [comment]
I'm sorry, maybe my question was confusing. I'm not looking to convert anything to JSON.

My understanding is that Swift 4 `Codable` allows for a unified interface that structs/classes can conform to to allow serialization.
Which format we encode `Codable` objects to is completely up to what `Encoder`/`Decoder` we implement, it's just that the only encoders included in `Foundation` are currently `PropertyListEncoder` and `JSONEncoder`. The Encoders aren't conforming to anything, it's just that they are built to accept objects that are `Codable` (=Serializable).
Both are really slow and bulky, and I'm currently looking for a smaller & faster format to serialize to.

I thought that maybe Protobuf could be a format that serializable objects could be serialized to.

**Comment by thomasvl** [comment]
Understood, some people (historically) have expected _numbers_ to always end up as _numbers_, and so the protobuf JSON isn't a match for them.  I'm not sure if supporting `Codable` would lead to any more of this.

Part of our stance has been if you want Protobufs, you want the protobuf behaviors, and it doesn't provide completely generic support for JSON encoding like some might expect.  Likewise, the keys in the JSON go through transforms from what is in the proto files, so that also can make it less of a fit for folks.

**Comment by allevato** [comment]
Supporting `Codable` would be somewhat complicated, because the protobuf spec describes multiple types that map into a single Swift type. For example, `uint64` and `fixed64` are both represented as `UInt64` in Swift, so we would need a way to discriminate them for serialization.

If we really wanted to do it (that's a big "if"), it's probably possible—generated messages could have their own `CodingKey` enum that tags each field with the additional information needed for each field (its field number and its wire format), and then the proto-specific encoders/decoders could pull that information and do the right thing. Whether that alone is enough is hard to say though, and there could be other gotchas.

However, it's extremely likely that such an implementation that has to go through all of the coding plumbing would be significantly slower than the implementation here, which has been specifically tuned for speed—and when using the binary encoding, speed is what you want.

In short, supporting the `Codable` APIs seem like a large amount of effort for relatively little gain. `Codable` is meant to allow arbitrary data structures to support serialization in multiple formats. Protocol buffers are specifically generated data structures that support serialization in a finite number of spec-defined formats. The use cases simply don't line up.

**Comment by tbkka** [comment]
> I could use a fast alternative to JSONEncoder or PropertyListEncoder...

I believe the performance of `JSONEncoder` and `PropertyListEncoder` is constrained primarily by the performance overhead of the `Codable` protocol itself.  SwiftProtobuf implementation included many iterations of profiling and redesigning the API between the generated objects and the encoders/decoders.  I'm not at all certain that adapting our encoders/decoders to work against the `Codable` protocols would provide any performance advantage.  `Codable` also lacks a number of key concepts that would be needed to properly support protobuf's binary serialization.  (For example, it has no concept of varying wire encodings for the same in-memory data type.)

**Comment by Bersaelor** [comment]
@allevato thank you, that made the differen use cases much clearer for me.

@tbkka Yes and no, from my profiling it seems the issue is as much with `Codable` as with the fact that `JSONCoder` uses `NSJSONSerialization` internally, which converts everything to `NSNumber ` and `NSString` etc. Similar issue for `PropertyListEncoder`.

**Comment by ephemer** [comment]
The concept of "varying wire encodings for the same in-memory data type" seems to have come up twice here as the only real reason not to implement `Codeable` conformance (I'm assuming this is the same thing as was meant with "uint64 and fixed64 are both represented as UInt64 in Swift"?)

I'm not sure I see how this is a hinderance to the goal as a level of consistency. The codec would indeed need a source of truth (defined externally to the `Codeable` type) about which protobuf tags to use, and which protobuf types map to which Swift types etc. But is it unreasonable to implement the codec such that it takes either a Swift data structure generated from a `.proto` file, or even the `.proto` file itself?

```
let decoder = ProtobufDecoder(schema: myRootSchema) // <-- here
let myInstance = try decoder.decode(MyType.self, from: binaryData)
```

Or is the criticism more of a "what do we gain for our efforts here?" I don't claim to be an expert at `Codeable`, and I'm especially not an expert in protobuf. Just trying to figure out why it'd be a non-goal to fit in with the forward momentum of Swift development in this regard. I was made interested in this idea after listening to a podcast with Chris Lattner recently, who seemed to suggest that `Codeable` compatibility for Protobuf was already available.

As a way of promoting use of of this library, it'd be pretty great to support this as a (admittedly slightly more complicated) drop-in addition to the existing JSON and PropertyList codecs.

**Comment by thomasvl** [comment]
So I'm starting to wonder if folks are talking about different things here.

I was thinking these requests were around supporting mapping the individual fields of a message to the encode/decode methods (int32, double, etc.).  So it would get another api to get the tree of JSON for example; i.e.- the library would be calling the encode/decode methods on a per field basis.  But now I'm starting to wonder if folks were thinking of having their Messages encoded as a single (binary) blob in the encoder/decoder and not having the fields expressed/exposed.  This would just allow a message to be captured within something Codable, but wouldn't expose the details of how it is encoded.

The comments about `JSONEncoder` or `PropertyListEncoder` performance; I guess folks could even be talking not so much about directly supporting `Codable` as much as trying to have this library expose things that conform the the Encoder/Decoder (and I guess the matching Containers), hoping that some how lets types that support Codable use our endcode/decode to get better performance?

Anyway, I guess I need folks commenting to better spell out what they mean by saying they want _Swift4 `Codable` support_ so we can be sure everyone is talking about the same thing.

Or it is just me confused and everyone can ignore this comment.  😄

**Comment by Bersaelor** [comment]
@ephemer 

> As a way of promoting use of of this library, it'd be pretty great to support this as a (admittedly slightly more complicated) drop-in addition to the existing JSON and PropertyList codecs.

This is exactly what I meant.

@thomasvl 
I was asking for protobuf to become an alternative to `JSONEncoder`/ `PropertyListEncoder` for serialization of objects conforming to `Codable`.

**Comment by tbkka** [comment]
> I was asking for protobuf to become an alternative to JSONEncoder/ PropertyListEncoder for serialization of objects conforming to Codable.

Do you want to exchange the serialized result with a server or other protobuf-using system?

* If so, then you have a `proto` file and you would have to find a way to ensure the result conformed to that schema (which has information that `Codable` by itself cannot represent).

* If not, then you might not need to conform to a `proto` file, but the result would not be protobuf and it would be misleading to call it that.

**Comment by thomasvl** [comment]
If the intent is just to avoid the performance of `JSONEncoder` / `PropertyListEncoder`, then it seems like the better thing to do would be to file bugs against those.  Protobuf can get some of its performance between we have a proto definition to provide the scheme and use a different set of interfaces.  It doesn't seem right to take on the burden of trying to support a different api just to work around someone's implementation speed (especially when some of the speed could be because of the api one is trying to duplicate).

**Comment by ephemer** [comment]
It’s not a complaint about performance of encoding or decoding JSON or plists so much as an interest in the benefit of what protobuf can give us. I’m interested in the safety guarantees and also size efficiency in storing and exchanging data as protobuf instead of in JSON. Or why else are we using protobuf at all??

As is, the data I’m interested in has (among other things) arrays that have hundreds of JSON objects with a few keys each. Their structure is such that using protobuf would allow a 90%+ OTW saving compared to the current JSON implementation. And yes, coding/decoding protobuf is faster too.

So, I could use Swift-protobuf as is, but the suggestion in this thread, as I understood it and continued, is that we make ProtobufDecoder / ProtobufEncoder types that work alongside JSONDecoder / JSONDecoder (etc) without having to think too much about new APIs. This would set a great precedent for Swift Codable types, given that Codable is now a first class citizen in Swift land.

**Comment by tbkka** [comment]
@ephemer:  As discussed earlier, the Codable protocol has some inherent limitations that make it very difficult to accurately support the official protobuf spec.  Since our primary goal is to accurately support that spec, providing a Codable-based encoder/decoder is not really appropriate _for the SwiftProtobuf project_.

**Comment by allevato** [comment]
It might be _technically_ possible for us to generate appropriate coding keys with the necessary supplemental information (field number and wire format) to let a hypothetical `ProtobufEncoder/Decoder` work, but without diving into the details, it's hard to say what stumbling blocks someone doing so would hit because of their different design goals. `Codable` is considerably more general than protobuf, so it's likely that there would be some degree of runtime checks to ensure that nobody used the protobuf encoders/decoders with incompatible data types. That kind of sucks. There may also just be places where the design principles between the two APIs can't fit together.

I'd still be concerned about performance, too. If a `Codable`-based implementation is significantly slower than the one we have now, that's kind of a deal breaker. There would be no benefit to offering and maintaining both a "full speed" binary encoder for users who need all the performance and a "kind of ok" binary encoder for users who just want to use a different API.

Since the protobuf spec defines a very specific JSON representation (e.g., the way 64-bit integers are handled), we would need to provide our own encoders/decoders for those as well, because the ones in Foundation wouldn't be appropriate. So that's awkward at best—if users want to communicate with a server that specifically uses JSON-encoded protobufs, they would need to use our encoder/decoder, not the standard one. And we would need to implement a Text Format encoder/decoder as well.

It could certainly be worth investigating, but what you're asking for is a complete redesign of the library internals and heavy code generation changes to fit a new API—that's **a lot** easier said than done. If somebody passionate about it wants to experiment with that and give us concrete data showing that all of the concerns stated above can be alleviated, feel free...

**Comment by ephemer** [comment]
@tbkka maybe I missed it, but I didn’t see any discussion following the suggestions I made to remedy the issues you mention, namely feeding the codegen’d Schema / proto file into the hypothetical `ProtobufDecoder(schema: xyz)` initialiser.

The Codable type itself doesn’t need to know anything about the protobuf implementation, only the Swift types. This would fit alongside the other Codable implementations perfectly, and makes more sense anyway given that the extra info needed by protobuf is only relevant to protobuf and not to other Codecs (JSON, plist etc).

If this seems like nonsense to you it’s possible we’re talking about different things: can you elaborate as to what you mean?

**Comment by ephemer** [comment]
@allevato thanks for your response, I’m not referring to JSON protobufs (nor was the OP from what I can tell), but rather the standard binary implementation.

Are you suggesting there are no runtime checks as to whether the data types are correct in Swift-Protobuf? I’m surprised by that revelation, but if it’s true you’re probably right that the implementations would differ enough to put this into the “too hard” basket.

**Comment by allevato** [comment]
I mentioned binary protobufs as well—we need to carry additional information than just the Swift type in order to encode it (field number and wire format), and our implementation is already highly tuned for performance because it's intentionally not generic in a lot of places and my concern is that a `Codable`-based implementation would have significant difficulty hitting those same numbers.

**Comment by AndyDentFree** [comment]
For future reference for anyone keen to implement a binary protocol for Codable, including deciding to map to Protobuf:

Mike Ash wrote a [great article about it on 2017-07-28](https://www.mikeash.com/pyblog/friday-qa-2017-07-28-a-binary-coder-for-swift.html) including [source on GitHub](https://github.com/mikeash/BinaryCoder)

I understand the appeal very strongly as a way for people to easily get into Protobuf but share concerns that it bypasses Protobuf's toolchain.

An interesting alternative would be using the Codable protocol to **generate** .proto files, then use the rest of the toolchain.

**Comment by cpatterson-lilly** [comment]
My $.02... This comment very succinctly seems to describe the basic problem with using protobufs in any problem domain:

> Codable is meant to allow arbitrary data structures to support serialization in multiple formats. Protocol buffers are specifically generated data structures that support serialization in a finite number of spec-defined formats. The use cases simply don't line up.

Having said that, it seem therefore reasonable to limit the use cases of protobufs with Codeables to protobuf's own use cases. In other words, if one were to try to serialize a Codeable object whose properties were not protobuf-compatible, the system should alert the programmer with a fatal error at build time.

Finally, it seems to me that this suggestion would be the best path forward:

> An interesting alternative would be using the Codable protocol to generate .proto files, then use the rest of the toolchain.

**Comment by AndyDentFree** [comment]
>Codable is meant to allow arbitrary data structures to support serialization in multiple formats.

Except that it doesn't really support arbitrary structures - it's only for flat structures, failing to support polymorphism via classes or protocols. Unless I have massively misunderstood something, you have to do a lot of work writing your own wrappers to get polymorphic decoding. It only recreates base classes.

I explored this a bit in [some playgrounds[(https://github.com/AndyDentFree/perswifting) ending up doing my own implementation to get easy complex graph support. 
(I **hate** wheel re-invention or any approximation to it, so would love someone to correct my misunderstandings.)

**Comment by GalCohen** [comment]
I think like many others up thread, the hope was that something like protobuf would come along and take our existing Codable models, which were up until now serialized to JSON, and instead use a binary format. Unfortunately, maybe protobufs isn’t a good fit for this use case.

**Comment by mickeyl** [comment]
FWIW, MsgPack (in particular, [this implementation](https://github.com/hirotakan/MessagePacker)) seems to be both faster (10%) and smaller (about 40%) than the `PropertyListEncoder`.

**Comment by thomasvl** [comment]
Looks like MessagePack follows a different wire specification, so it isn't an exact match.

**Comment by mickeyl** [comment]
Yes, this comment was written based on the premise that protobuf and Codable is not a good match due to the dynamic/static mismatch.

**Comment by tbkka** [comment]
I'm inclined to close this.

To summarize:  There are a few key obstacles for this idea, fundamental mismatches between how Codable works and how Protobuf works:
* Protobuf serialization records field _numbers_.  (Indeed, this is a major part of protobuf's size efficiency.)  Swift Codable has no such concept; it traffics in field _names_.
* Protobuf has multiple wire formats for each Swift data type.  For example, protobuf has several different "int32" encodings that are optimized for values that are mostly positive, mostly small, or often large.  Codable has no such concept.

To resolve the above issues, there aren't a lot of options:
* You could generate the requisite information automatically. (e.g., number fields in definition order, choose a single encoding for each swift format). But this would preclude interoperating with other protobuf-using systems, which have that information explicitly specified in a .proto file.  I think of this as "not really protobuf."
* You could work out some way to augment the Swift definition (using macros, perhaps).  This would require every data type used by your system to be explicitly annotated for use with this coder, which would make the coder useless with other Swift data types that supported Codable.  I think of this as "not really Codable."

I think it would be quite interesting for someone to start a new project that used protobuf concepts to develop a new encoder that works with Codable.  You're even welcome to copy code from SwiftProtobuf (within the restrictions of our Copyright License, of course).  But I still see no point in _the SwiftProtobuf project_ attempting this.
